fxrunner/src/lib/request.rs

```diff
@@ -0,0 +1,176 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::borrow::Cow;
+use std::io;
+use std::path::{Path, PathBuf};
+
+use async_trait::async_trait;
+use rand::prelude::*;
+use thiserror::Error;
+use tokio::fs::create_dir;
+
+use crate::fs::PathExt;
+
+#[derive(Clone)]
+pub struct RequestInfo<'a> {
+    pub id: Cow<'a, str>,
+    pub path: PathBuf,
+}
+/// A trait for creating and validating requests.
+#[async_trait]
+pub trait RequestManager {
+    /// Create a new request.
+    async fn new_request(&self) -> Result<RequestInfo<'static>, NewRequestError>;
+
+    /// Attempt to resume a request with the given ID.
+    ///
+    /// If the request ID is valid and a prior request exists, the path to its
+    /// directory will be returned.
+    async fn resume_request<'a>(
+        &self,
+        request_id: &'a str,
+    ) -> Result<RequestInfo<'a>, ResumeRequestError>;
+
+    /// Ensure the profile directory for the given request exists.
+    async fn ensure_profile_dir<'a>(
+        &self,
+        request_info: &RequestInfo<'a>,
+    ) -> Result<PathBuf, io::Error>;
+}
+
+pub struct FsRequestManager {
+    path: PathBuf,
+}
+
+impl FsRequestManager {
+    pub fn new(path: &Path) -> Self {
+        FsRequestManager { path: path.into() }
+    }
+}
+
+#[async_trait]
+impl RequestManager for FsRequestManager {
+    async fn new_request(&self) -> Result<RequestInfo<'static>, NewRequestError> {
```

Comment:
I think the implementation of `new_request` needs just a little bit of commenting explaining why we're doing what we're doing.